cpp/src/arrow/compute/kernels/scalar_cast_string.cc

```diff
@@ -327,31 +329,43 @@ BinaryToBinaryCastExec(KernelContext* ctx, const ExecSpan& batch, ExecResult* ou
     }
   }
 
-  const int64_t sum_of_binary_view_sizes = util::SumOfBinaryViewSizes(
-      input.GetValues<BinaryViewType::c_type>(1), input.length);
-
-  // TODO(GH-43573): A more efficient implementation that copies the validity
-  // bitmap all at once is possible, but would mean we don't delegate all the
-  // building logic to the ArrayBuilder implementation for the output type.
-  OutputBuilderType builder(options.to_type.GetSharedPtr(), ctx->memory_pool());
-  RETURN_NOT_OK(builder.Resize(input.length));
-  RETURN_NOT_OK(builder.ReserveData(sum_of_binary_view_sizes));
-  arrow::internal::ArraySpanInlineVisitor<I> visitor;
-  RETURN_NOT_OK(visitor.VisitStatus(
-      input,
-      [&](std::string_view v) {
-        // Append valid string view
-        return builder.Append(v);
+  ArrayData* output = out->array_data().get();
+  output->length = input.length;
+  output->SetNullCount(input.null_count);
+
+  // Set up bitmap
+  if (input.offset == output->offset) {
+    output->buffers[0] = input.GetBuffer(0);
+  } else {
+    if (input.buffers[0].data != NULLPTR) {
+      ARROW_ASSIGN_OR_RAISE(
+          output->buffers[0],
+          arrow::internal::CopyBitmap(ctx->memory_pool(), input.buffers[0].data,
+                                      input.offset, input.length));
+    }
+  }
+
+  // Set up offset and data buffer
+  DataBuilder data_builder(ctx->memory_pool());
```

Comment:
The previous implementation caculate a  `sum_of_binary_view_sizes`, and `ReserveData` for it. Why did here doesn't use same way to reserve data? Would blindly check append the buffer being faster?

Reply:
Sorry I missed this optimization in the original code. Just added it back.